---
title: "scRNAseq_kategtf_aligned"
author: "Kate Randall"
date: "2025-05-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages


```{r}
library(plyr)
library(tidyverse)
library(svglite)
library(ggplot2)
library(ggpackets)
library(patchwork)
library(glmGamPoi)
library(presto)
library(BPCells)
library(RANN)
library(Rtsne)
library(irlba)
library(png)
library(qqconf)
library(lmtest)
library(metap)
library(multtest)
library(mime)
library(ellipsis)
library(lazyeval)
library(stringi)
library(reshape2)
library(polyclip)
library(goftest)
library(RColorBrewer)
library(viridis)
library(beepr)
library(SeuratObject)
library(Seurat)
#library(ggseurat)
library(ggpackets)
library(sctransform)
library(scCustomize)

# set seed for reproducibility
set.seed(123456)

# set assay version to Seurat v3
#options(Seurat.object.assay.version = "v3")

```


# Import scRNA-seq dataset


```{Import 10X scRNA-seq dataset}

# 0hr timepoint
hr0.data <- Read10X(data.dir = "/home/bret/scRNAseq_kategtf_cellranger/head_regen/outs/raw_feature_bc_matrix")

hr0 <- CreateSeuratObject(counts = hr0.data, assay = "RNA")
class(hr0[["RNA"]])

# add metadata column "timepoint"
hr0[["timepoint"]] <- "hr0"

# check metadata
hr0_metadata <- hr0[[]]
(colnames(hr0_metadata))
head(hr0_metadata)

# 3hr timepoint
hr3.data <- Read10X(data.dir = "/home/bret/scRNAseq_kategtf_cellranger/head_regen3hrs/outs/raw_feature_bc_matrix")
hr3 <- CreateSeuratObject(counts = hr3.data)
class(hr3[["RNA"]])
hr3[["timepoint"]] <- "hr3"

# 6 hr timepoint
hr6.data <- Read10X(data.dir = "/home/bret/scRNAseq_kategtf_cellranger/head_regen6hrs/outs/raw_feature_bc_matrix")
hr6 <- CreateSeuratObject(counts = hr6.data)
class(hr6[["RNA"]])
hr6[["timepoint"]] <- "hr6"

# 12hr timepoint
hr12.data <- Read10X(data.dir = "/home/bret/scRNAseq_kategtf_cellranger/head_regen12hrs/outs/raw_feature_bc_matrix")
hr12 <- CreateSeuratObject(counts = hr12.data)
class(hr12[["RNA"]])
hr12[["timepoint"]] <- "hr12"

# 18hr timepoint
hr18.data <- Read10X(data.dir = "/home/bret/scRNAseq_kategtf_cellranger/head_regen18hrs/outs/raw_feature_bc_matrix")
hr18 <- CreateSeuratObject(counts = hr18.data)
class(hr18[["RNA"]])
hr18[["timepoint"]] <- "hr18"

# 24hr timepoint
hr24.data <- Read10X(data.dir = "/home/bret/scRNAseq_kategtf_cellranger/head_regen24hrs/outs/raw_feature_bc_matrix")
hr24 <- CreateSeuratObject(counts = hr24.data)
class(hr24[["RNA"]])
hr24[["timepoint"]] <- "hr24"

# 36hr timepoint
hr36.data <- Read10X(data.dir = "//home/bret/scRNAseq_kategtf_cellranger/head_regen36hrs/outs/raw_feature_bc_matrix")
hr36 <- CreateSeuratObject(counts = hr36.data)
class(hr36[["RNA"]])
hr36[["timepoint"]] <- "hr36"

# SAVE & LOAD
save(hr0, hr3, hr6, hr12, hr18, hr24, hr36, file="hr.meta.kg.RData")
#load(file = "hr.meta.kg.RData")

```


# Custom plot theme


```{r}

krplot <- theme(text=element_text(size=18, color="black"),
        rect=(element_blank()),
        axis.line=element_line(),
        axis.text=element_text(size=14, color="black"),
        plot.title=element_text(hjust=0.5, face = "bold", size = 22))

```



# Quality Control Metrics


## Check # cells

```{r}
# function to return # cells in Seurat object
get_cell_count <- function(count_this) {
  return(ncol(count_this))
}

# usage
#count_this <- hr.merged.kg
#cell_count <- get_cell_count(count_this)
#cat("Number of cells in the Seurat object:", cell_count, "\n")

# check # cells raw counts
count_this <- hr0
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr0:", cell_count, "\n")

count_this <- hr3
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr3:", cell_count, "\n")

count_this <- hr6
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr6:", cell_count, "\n")

count_this <- hr12
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr12:", cell_count, "\n")

count_this <- hr18
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr18:", cell_count, "\n")

count_this <- hr24
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr24:", cell_count, "\n")

count_this <- hr36
cell_count <- get_cell_count(count_this)
cat("Number of cells in hr36:", cell_count, "\n")

```



## Merge timepoints (for QC only)

```{Merge timepoints - QUICKLY with multicore!}

# in Terminal window!

R

# load packages and inputs as above
library(future)
library(Seurat)

load(file = "hr.meta.kg.RData")


# REMOVE LATER force v3 assay (instead of v5)
#options(Seurat.object.assay.version = "v3")

# set up for parallelization 
plan("multicore", workers = parallel::detectCores())
options(future.globals.maxSize = 100 * 1024^3) # use <100 GB RAM

# log compute time
start_time <- Sys.time()

# create merged object (multicore)
hr.merged.kg <- value(future({
  merge(x = hr0,
        y = list(hr3, hr6, hr12, hr18, hr24, hr36),
        add.cell.ids = c("0hrs", "3hrs", "6hrs", "12hrs", "18hrs", "24hrs", "36hrs"))
}))

save(hr.merged.kg, file = "hr.merged.kg.RData")

# report compute time
end_time <- Sys.time()
cat("Merge compute time:", end_time - start_time)

```

## Raw data wrangling

```{r}
# load back into RStudio
load(file = "hr.merged.kg.RData")

count_this <- hr.merged.kg
cell_count <- get_cell_count(count_this)
cat("Number of cells in the Seurat object:", cell_count, "\n")


# clean up env
rm(hr0_metadata, hr0.data, hr3.data, hr6.data, hr12.data, hr18.data, hr24.data, hr36.data)


raw.metadata <- hr.merged.kg@meta.data
unique(raw.metadata$timepoint)

#force chronological timepoint order (default is alphanumeric)
raw.metadata$timepoint <- factor(raw.metadata$timepoint,
                             levels = c("hr0", "hr3", "hr6", "hr12", "hr18", "hr24", "hr36"))

# copy cell ids into new column "timepoint"
hr.merged.kg[["timepoint"]] <- sapply(Cells(hr.merged.kg), function(x) strsplit(x, "_")[[1]][[1]])
unique(hr.merged.kg@meta.data$timepoint)


# rename columns 
raw.metadata <- hr.merged.kg@meta.data
raw.metadata$cells <- rownames(raw.metadata)
raw.metadata <- raw.metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr.merged.kg@meta.data <- raw.metadata
head(hr.merged.kg@meta.data)


# SAVE & LOAD          
save(hr.merged.kg, file = "hr.merged.kg.RData")
#load(file = "hr.merged.kg.RData")
```


## View raw QC data

```{r}

######################## nCells, nUMI (aka nCount), & nGene (aka nFeature) #################################
 
#raw data before normalization, scaling, or filtration!

# plot (bar) nCells per timepoint
nCells_raw <- raw.metadata %>%
  ggplot(aes(x= timepoint, fill= timepoint)) +
  geom_bar() +
  krplot +
  scale_y_continuous(expand=(c(0,0))) +
  scale_fill_viridis_d() +
  ggtitle("nCells (raw)")
nCells_raw
ggsave("nCells_raw", plot = nCells_raw, device = "pdf")


# plot nUMIs per cell (density)
UMIdensity_raw <- raw.metadata %>% 
  ggplot(aes(x=nUMI, fill= timepoint, group = timepoint)) + 
  geom_density(alpha = 0.4, fill = "gray10") + 
  krplot +
  scale_x_continuous(labels = scales::comma, 
                     trans = "log10", 
                     limits = c(1,100000), 
                     n.breaks = 8) +
  scale_y_continuous(expand=(c(0,0))) +
  ylab("log10 cell density") +
  ggtitle("log10 UMIs per cell")+
  geom_vline(xintercept = 100, color = "red")
UMIdensity_raw
ggsave("UMI_density_raw", plot = UMIdensity_raw, device = "pdf")



# plot nGenes per cell (density)
genedensity_raw <- raw.metadata %>% 
  ggplot(aes(x=nGene, fill= timepoint, group = timepoint)) + 
  geom_density(alpha = 0.4, fill = "gray10") +
  krplot +
  scale_x_continuous(labels = scales::comma, 
                     trans = "log10", 
                     limits = c(1,100000),
                     expand=c(0,0)) +
  scale_y_continuous(limits = c(0,1.4),
                     expand=c(0,0)) +
  ggtitle("nGenes per cell") +
  # + geom_vline(xintercept = 200) +
  ylab("log10 cell density")
genedensity_raw
ggsave("nGenepercell_raw", plot = genedensity_raw, device = "pdf")


# plot nGene vs nUMI (scatter)
genexumi_raw <- raw.metadata %>% 
  ggplot(aes(x=nUMI, y=nGene)) + 
  geom_point(alpha = 0.3) + 
  krplot +
  stat_smooth(method=lm) +
  scale_x_log10(labels = scales::comma,
                     n.breaks = 8,
                     limits = c(1,100000),
                     expand=c(0,0)) +
  scale_y_log10(labels = scales::comma,
                expand=c(0,0)) +
  ggtitle("nGene per nUMI")
genexumi_raw
ggsave("genexumi_raw", plot = genexumi_raw, device = "pdf")

######################## Complexity scores #################################

# Complexity score (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
raw.metadata$log10GenesPerUMI <- log10(hr.merged.kg$nGene) / log10(hr.merged.kg$nUMI)
hr.merged.kg$log10GenesPerUMI <- log10(hr.merged.kg$nGene) / log10(hr.merged.kg$nUMI)


# complexity scores (# genes detected per UMI)

complexity <- raw.metadata %>%
  ggplot(aes(x=log10GenesPerUMI, fill=timepoint)) +
  geom_density(alpha = 0.4) +
  krplot +
  scale_x_continuous(expand=c(0,0))+
  scale_y_continuous(expand=c(0,0)) +
  ggtitle("Complexity score") +
  geom_vline(xintercept = 0.8) +
  ylab("Cell density")
complexity
ggsave("complexity", plot = complexity, device = "pdf")

```

## Remove empty droplets, first QC filter,  & repeat QC plots

```{r}
# remove empty droplets
hr.merged.kg <- subset(hr.merged.kg, subset = nUMI > 200)

# remove low quality cells
hr.merged.kg <- subset(hr.merged.kg, subset = nGene > 200)


metadata <- hr.merged.kg@meta.data


metadata$timepoint <- factor(metadata$timepoint,
                             levels = c("0hrs", "3hrs", "6hrs", "12hrs", "18hrs", "24hrs", "36hrs"))

save(hr.merged.kg, metadata, file = "hr.merged.kg.droplets")

#load(file= "hr.merged.kg.droplets.RData")


# run Cellbender (see python scripts) and import results
# ERROR: cant pickle weakrefs 

# compare Cellbender results to nCount_RNA filter
# would go here if I could get Cellbender to work :(




######################## nCells, nUMI (aka nCount), & nGene (aka nFeature) #################################
 
# after empty droplets removed, before normalization, scaling, or filtration!

# plot (bar) nCells per timepoint
nCells <- metadata %>%
  ggplot(aes(x= timepoint, fill= timepoint)) +
  geom_bar() +
  krplot +
  scale_y_continuous(labels = scales::label_comma(),
                     limits = c(0,80000),
                     n.breaks = 7) +
  scale_fill_viridis_d() +
  ggtitle("nCells")
nCells
ggsave("nCells_emptydropletsremoved", plot = nCells, device = "pdf")


# plot nUMIs per cell (density)
UMIdensity <- metadata %>% 
  ggplot(aes(x=nUMI, fill= timepoint, group = timepoint)) + 
  geom_density(alpha = 0.4, fill = "gray10") + 
  krplot +
  scale_x_continuous(labels = scales::label_comma(),
                     limits = c(0,25000)) +
  scale_y_continuous(expand=c(0,0)) +
  ylab("cell density") +
  ggtitle("UMIs per cell") +
  geom_vline(xintercept = 100, color = "red")
UMIdensity
ggsave("UMI_density_emptydropletsremoved", plot = UMIdensity, device = "pdf")



# plot nGenes per cell (density)
genedensity <- metadata %>% 
  ggplot(aes(x=nGene, fill= timepoint, group = timepoint)) + 
  geom_density(alpha = 0.4, fill = "gray10") +
  krplot +
  scale_x_continuous(labels = scales::comma, 
                     trans = "log10", 
                     limits = c(10,10000), 
                     n.breaks = 8,
                     expand=c(0,0)) +
  scale_y_continuous(limits = c(0,4),
                     n.breaks = 8,
                     expand=c(0,0)) +
  ggtitle("nGenes per cell") +
  # + geom_vline(xintercept = 250) +
  ylab("log10 cell density")
genedensity
ggsave("nGenepercell_emptydropletsremoved", plot = genedensity, device = "pdf")


# plot nGene vs nUMI (scatter)
genexumi <- metadata %>% 
  ggplot(aes(x=nUMI, y=nGene)) + 
  geom_point(alpha = 0.3) + 
  krplot +
  stat_smooth(method=lm) +
  scale_x_log10(labels = scales::comma,
                     n.breaks = 8,
                     limits = c(650,100000),
                     expand=c(0,0)) +
  scale_y_log10(labels = scales::comma,
                expand=c(0,0)) +
  ggtitle("nGene per nUMI")
genexumi
ggsave("genexumi", plot = genexumi, device = "pdf")


# complexity scores (# genes detected per UMI)
complexity <- metadata %>%
  ggplot(aes(x=log10GenesPerUMI, fill=timepoint)) +
  geom_density(alpha = 0.4) +
  krplot +
  scale_x_continuous(expand=c(0,0))+
  scale_y_continuous(expand=c(0,0)) +
  ggtitle("Complexity score") +
  #geom_vline(xintercept = 0.8) +
  ylab("Cell density")
complexity
ggsave("complexity", plot = complexity, device = "pdf")

```


# QC Filtering (each timepoint individually)


```{r}

#load objects
load(file="hr.meta.kg.RData")


#Prepare each Seurat object
## hr0 ##
#view merged metatdata
View(hr0@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr0$log10GenesPerUMI <- log10(hr0$nFeature_RNA) / log10(hr0$nCount_RNA)
#rename columns
metadata <- hr0@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr0@meta.data <- metadata
View(hr0@meta.data)

## hr3 ##
#view merged metatdata
View(hr3@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr3$log10GenesPerUMI <- log10(hr3$nFeature_RNA) / log10(hr3$nCount_RNA)
#rename columns
metadata <- hr3@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr3@meta.data <- metadata

## hr6 ##
#view merged metatdata
View(hr6@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr6$log10GenesPerUMI <- log10(hr6$nFeature_RNA) / log10(hr6$nCount_RNA)
#rename columns
metadata <- hr6@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr6@meta.data <- metadata

## hr12 ##
#view merged metatdata
View(hr12@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr12$log10GenesPerUMI <- log10(hr12$nFeature_RNA) / log10(hr12$nCount_RNA)
#rename columns
metadata <- hr12@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr12@meta.data <- metadata

## hr18 ##
#view merged metatdata
View(hr18@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr18$log10GenesPerUMI <- log10(hr18$nFeature_RNA) / log10(hr18$nCount_RNA)
#rename columns
metadata <- hr18@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr18@meta.data <- metadata

## hr24 ##
#view merged metatdata
View(hr24@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr24$log10GenesPerUMI <- log10(hr24$nFeature_RNA) / log10(hr24$nCount_RNA)
#rename columns
metadata <- hr24@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr24@meta.data <- metadata

## hr36 ##
#view merged metatdata
View(hr36@meta.data)
# Compute novelty score and add column (logGenesPerUMI = logGenesPerCell / logUMIsPerCell)
hr36$log10GenesPerUMI <- log10(hr36$nFeature_RNA) / log10(hr36$nCount_RNA)
#rename columns
metadata <- hr36@meta.data
metadata$cells <- rownames(metadata)
metadata <- metadata %>%
  dplyr::rename(nUMI = nCount_RNA,
                nGene = nFeature_RNA)
hr36@meta.data <- metadata


##Filter each timepoint independently
hr0f <- subset(hr0, subset = (nUMI >= 650) & 
                 (nGene >= 250) & 
                 (log10GenesPerUMI > 0.8))

hr3f <- subset(hr3, subset = (nUMI >= 650) & 
                 (nGene >= 250) & 
                 (log10GenesPerUMI > 0.8))

hr6f <- subset(hr6, subset = (nUMI >= 650) & 
                 (nGene >= 250) & 
                 (log10GenesPerUMI > 0.8))

hr12f <- subset(hr12, subset = (nUMI >= 650) & 
                  (nGene >= 250) & 
                  (log10GenesPerUMI > 0.8))

hr18f <- subset(hr18, subset = (nUMI >= 650) & 
                  (nGene >= 250) & 
                  (log10GenesPerUMI > 0.8))

hr24f <- subset(hr24, subset = (nUMI >= 650) & 
                  (nGene >= 250) & 
                  (log10GenesPerUMI > 0.8))

hr36f <- subset(hr36, subset = (nUMI >= 650) & 
                  (nGene >= 250) & 
                  (log10GenesPerUMI > 0.8))

#Scatter plots
#hr0
hr0scatter <- FeatureScatter(hr0, feature1 = "nUMI", feature2 = "nGene")
hr0fscatter <- FeatureScatter(hr0f, feature1 = "nUMI", feature2 = "nGene")
svglite("hr0filter_scatter.svg")
hr0scatter | hr0fscatter
dev.off()

#hr3
hr3scatter <- FeatureScatter(hr3, feature1 = "nUMI", feature2 = "nGene")
hr3fscatter <- FeatureScatter(hr3f, feature1 = "nUMI", feature2 = "nGene")
hr3scatter | hr3fscatter

#hr6
hr6scatter <- FeatureScatter(hr6, feature1 = "nUMI", feature2 = "nGene")
hr6fscatter <- FeatureScatter(hr6f,feature1 = "nUMI", feature2 = "nGene")
hr6scatter | hr6fscatter

#hr12
hr12scatter <- FeatureScatter(hr12, feature1 = "nUMI", feature2 = "nGene")
hr12fscatter <- FeatureScatter(hr12f, feature1 = "nUMI", feature2 = "nGene")
hr12scatter | hr12fscatter

#hr18
hr18scatter <- FeatureScatter(hr18, feature1 = "nUMI", feature2 = "nGene")
hr18fscatter <- FeatureScatter(hr18f, feature1 = "nUMI", feature2 = "nGene")
hr18scatter | hr18fscatter

#hr24
hr24scatter <- FeatureScatter(hr24, feature1 = "nUMI", feature2 = "nGene")
hr24fscatter <- FeatureScatter(hr24f, feature1 = "nUMI", feature2 = "nGene")
hr24scatter | hr24fscatter

#hr36
hr36scatter <- FeatureScatter(hr36, feature1 = "nUMI", feature2 = "nGene")
hr36fscatter <- FeatureScatter(hr36f, feature1 = "nUMI", feature2 = "nGene")
hr36scatter | hr36fscatter

#Filter hr.merged (for visualization purposes only)
hr.merged.kg.f <- subset(hr.merged.kg, subset = (nUMI >= 650) & 
                        (nGene >= 250) & 
                        (log10GenesPerUMI > 0.8)
                       )
hr.merged.scatter <- FeatureScatter(hr.merged.kg, feature1 = "nUMI", feature2 = "nGene")
hr.merged.fscatter <- FeatureScatter(hr.merged.kg.f, feature1 = "nUMI", feature2 = "nGene")
svglite("hr.mergedfilter_scatter.svg")
hr.merged.scatter | hr.merged.fscatter
dev.off()

#plot n cells per timepoint before & after filtering
#Plot nCells per timepoint
n <- metadata %>%
  ggplot(aes(x=timepoint, fill=timepoint)) +
  geom_bar(aes(x = factor(timepoint), levelorder)) +
  #theme() +
  scale_y_continuous(expand=c(0,0)) +
  ggtitle("nCells per timepoint")

#custom palette RdYlBu in chronological order
chronRdYlBu_palette <- c("#d73027",
                         "#ffffbf",
                         "#e0f3f8",
                         "#91bfdb",
                         "#fc8d59",
                         "#4575b4",
                         "#fee090")

n + scale_fill_manual(values = chronRdYlBu_palette) + DarkTheme() + guides(fill="none")
ggsave("nCells_before_after_filtering", plot = n, device = "pdf")

nf <- hr.merged.kg.f %>%
  ggplot(aes(x=timepoint, fill=timepoint)) +
  geom_bar(aes(x = factor(timepoint, level = timepoint))) +
  theme() +
  scale_y_continuous(expand=c(0,0)) +
  ggtitle("nCells per timepoint")
#custom palette RdYlBu in chronological order
chronRdYlBu_palette <- c("#d73027",
                         "#ffffbf",
                         "#e0f3f8",
                         "#91bfdb",
                         "#fc8d59",
                         "#4575b4",
                         "#fee090")

nf + scale_fill_manual(values = chronRdYlBu_palette) + DarkTheme() + guides(fill="none")
ggsave("nCellsAfterFiltering", plot = nf, device = "pdf")


#SAVE AND LOAD
save(hr0f, hr12f, hr18f, hr24f, hr36f, hr3f, hr36f, hr6f, hr.merged.f, file="hr.meta.f.RData")
#load(file = "hr.meta.f.RData)
#set.seed(123456)


########
################# SCTransform individual timepoints #################
########

#Normalize, Scale, and find VariableFeatures for each timepoint independently
hr0fs <- SCTransform(hr0f, vst.flavor = "v2", variable.features.n = 3000)
hr3fs <- SCTransform(hr3f, vst.flavor = "v2", variable.features.n = 3000)
hr6fs <- SCTransform(hr6f, vst.flavor = "v2", variable.features.n = 3000)
hr12fs <- SCTransform(hr12f, vst.flavor = "v2", variable.features.n = 3000)
hr18fs <- SCTransform(hr18f, vst.flavor = "v2", variable.features.n = 3000)
hr24fs <- SCTransform(hr24f, vst.flavor = "v2", variable.features.n = 3000)
hr36fs <- SCTransform(hr36f, vst.flavor = "v2", variable.features.n = 3000)

hrfs.list <- c(hr0fs, hr3fs, hr6fs, hr12fs, hr18fs, hr24fs, hr36fs)


#SAVE AND LOAD
save(hr0fs, hr12fs, hr18fs, hr24fs, hr36fs, hr3fs, hr36fs, hr6fs, hrfs.list, file="hr.meta.f.SCT.RData")
```


## TIMEPOINT INTEGRATION

```{r}
#load(file = "hr.meta.f.SCT.RData")
#set.seed(123456)


#Select features that are repeatedly variable across datasets for integration
hrfs.var_features <- SelectIntegrationFeatures(object.list = hrfs.list, nfeatures = 3000)

hrfs.list <- PrepSCTIntegration(hrfs.list, anchor.features = hrfs.var_features)

#Identify Anchors
hrfs.anchors <- FindIntegrationAnchors(object.list = hrfs.list, 
                                       normalization.method = "SCT", 
                                       anchor.features = hrfs.var_features)


#Integration
hr.int <- IntegrateData(anchorset = hrfs.anchors, normalization.method = "SCT")


#Specify that downstream analysis is on integrated dataset
DefaultAssay(hr.int) <- "integrated"

#Regroup idents based on timepoint
RegroupIdents(hr.int, "timepoint")


#Reorder timepoints in chronological order
hr.int$timepoint <- factor(hr.int$timepoint, 
                           levels = c("hr0", "hr3", "hr6", "hr12", "hr18", "hr24", "hr36"))

unique(sapply(X = strsplit(colnames(hr.int), split = "_"), FUN = "[", 1))


#Scale and center data
ScaleData(hr.int,
          model.use = "linear",
          use.umi = FALSE,
          do.scale = TRUE,
          do.center = TRUE)


#SAVE AND LOAD
save(hr.int, file="hr.int.RData")
```


## PCA & Clustering

```{r}
#load(file = "hr.int.RData")
#set.seed(123456)


hr.int <- RunPCA(hr.int, npcs = 100)

#Plot PCA
VizDimLoadings(hr.int, 
               dims = 1:10, 
               reduction = "pca")

DimPlot(hr.int, 
        reduction = "pca",
        pt.size = 0.3)

DimHeamap(hr.int, dims = 1, 
           cells = 500, 
           balanced = TRUE)

DimHeatmap(hr.int, dims = 1:10,
           cells = 500, 
           balanced = TRUE)


#Elbow Plot
ElbowPlot(hr.int, ndims = 50)

#Elbow Plot quantification
pct <- hr.int[["pca"]]@stdev / sum(hr.int[["pca"]]@stdev) * 100
cumu <- cumsum(pct)
co1 <- which(cumu > 90 & pct <5) [1]

co2 <- sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) < 0.1), decreasing = T) [1] + 1
co2

pcs <- min(co1,co2)
pcs

plot_df <- data.frame(pct = pct,
                      cumu = cumu,
                      rank = 1:length(pct))
ggplot(plot_df, aes(cumu, pct, label = rank, color = rank>pcs)) +
  geom_text() +
  geom_vline(xintercept = 90, color = "grey") +
  geom_hline(yintercept = min(pct[pct>5]), color = "grey")+
  theme_bw()

#SAVE AND LOAD
save(hr.int, file="hr.int.RData")
#load(file = "hr.int.RData")
#set.seed(123456)


########
################# Clustering #################
########

#Run UMAP
hr.int <- RunUMAP(hr.int, 
                  reduction = "pca", 
                  dims = 1:20)

# Find Neighbors
hr.int <- FindNeighbors(hr.int, 
                        reduction = "pca", 
                        dims = 1:20)

#Set Clustering Resolution 
hr.int <- FindClusters(hr.int, resolution = 0.2)

#Plot UMAP (by clusters)
UMAP.cluster <- DimPlot(hr.int, 
                        reduction = "umap", 
                        group.by = "seurat_clusters", 
                        label = TRUE, 
                        repel = FALSE,
                        pt.size = 0.4)
UMAP.cluster




#FeaturePlot(hr.int, features = c(hr.int@timepoint("hr0", "hr36"), split.by = "timepoint")


#Tabulate nCells per cluster by timepoint
n_cells <- FetchData(hr.int,
                     vars = c("timepoint", "ident")) %>%
  dplyr::count(timepoint, ident) %>%
  tidyr::spread(ident, n)
View(n_cells)
write.csv(n_cells, file = "hr.int.nCellspercluster.bytimepoint.csv", row.names = TRUE)

#Plot nCells per timepoint
nc <- FetchData(hr.int, vars = c("timepoint", "seurat_clusters"))
ncells.clustxtime <- ggplot(nc, mapping = aes(x = seurat_clusters, fill= timepoint))
ncells.clustxtime + 
  geom_bar() + 
  scale_fill_brewer(palette="RdYlBu")
ggsave("nCells.clustxtime", plot = ncells.clustxtime, device = "pdf")

# plot (bar) nCells per timepoint
nc <- FetchData(hr.int, vars = c("timepoint"))

# Plot total cells per timepoint (single bar per timepoint)
ncells.timepoint <- ggplot(nc, aes(x = timepoint, fill = timepoint)) +
  geom_bar() +
  theme() +
  scale_fill_viridis_d() +
  scale_y_continuous(labels = scales::label_comma()) +
  ggtitle("Total cells per timepoint")
ncells.timepoint
ggsave("nCells_per_timepoint.pdf", plot = ncells.timepoint, device = "pdf")


########
################# Marker Gene Expression Plotting #################
########

##Piwi-1 
#Find numeric value of 10th quantile to use as NA cutoff
piwi.exp <- FetchData(object = hr.int, vars = "h1SMcG0013999")
q10 <- quantile(piwi.exp, na.rm = TRUE, probs = 0.90)
#plot exp on UMAP
FeaturePlot_scCustom(seurat_object = hr.int,
            features = "h1SMcG0013999", #piwi-1
            pt.size = 0.4, 
            order = TRUE,
            na_cutoff = q10,
            colors_use = viridis_dark_high,
            na_color = "gray")

#pax6a
#Find numeric value of 2nd quantile to use as NA cutoff
pax6a.exp <- FetchData(object = hr.int, vars = "h1SMcG0018062")
q10 <- quantile(pax6a.exp, na.rm = TRUE, probs = 0.90)
#plot exp on UMAP
FeaturePlot_scCustom(seurat_object = hr.int,
                     features = "h1SMcG0018062",
                     pt.size = 0.4, 
                     order = TRUE,
                     na_cutoff = q10,
                     colors_use = viridis_dark_high,
                     na_color = "gray")

FeaturePlot(hr.int, "h1SMcG0018062", pt.size = 1, order = TRUE, min.cutoff = 'q10') #pax6a



